homeassistant/helpers/entity.py

```diff
@@ -340,26 +412,41 @@ def _device_class_name(self) -> str | None:
         )
         return self.platform.component_translations.get(name_translation_key)
 
+    def __name_translation_key(self) -> str | None:
+        """Return translation key for entity name."""
+        if self.translation_key is None:
+            return None
+        assert self.platform
+        return (
+            f"component.{self.platform.platform_name}.entity.{self.platform.domain}"
+            f".{self.translation_key}.name"
+        )
+
     @property
     def name(self) -> str | None:
-        """Return the name of the entity."""
+        """Return the name of the entity.
+
+        Returns None if the name is set to DEVICE_NAME.
+        """
         if hasattr(self, "_attr_name"):
-            if self._attr_name is DEVICE_CLASS_NAME:
+            if (name := self._attr_name) is DEVICE_CLASS_NAME:
                 return self._device_class_name()
-            return self._attr_name
-        if self.translation_key is not None and self.has_entity_name:
+            if name is DEVICE_NAME:
```

Comment:
I'd rather use a sentinel object as default and keep the meaning that `None` is returned when we don't want an entity name. This would remove the implicit behavior while avoiding introducing a new constant requiring learning a new constant.

Reply:
Implemented that way in https://github.com/home-assistant/core/pull/94574